### PR TITLE
GDScript: Partially allow member lookup on invalid scripts

### LINKED ISSUE #92610: Invalid call. Nonexistent function 'reload' in base 'GDScript'.
### Tested versions

v4.3.beta1.mono.official [a4f2ea91a]

### System information

Godot v4.3.beta1.mono - Windows 10.0.17763 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2060 (NVIDIA; 31.0.15.2849) - AMD Ryzen 9 5950X 16-Core Processor (32 Threads)

### Issue description

Coding like this:
```
extends Node

var obj = RefCounted.new()
var gds = GDScript.new()

func _ready():
	gds.source_code = "func a():return 100"
	gds.reload()
	obj.set_script(gds)
	var rs = obj.a()
	prints(rs)

```

image
This code work before 4.3dev6 and output 100.

### Steps to reproduce

run the MRP, it report error

### Minimal reproduction project (MRP)

[gdstest.zip](https://github.com/user-attachments/files/15519573/gdstest.zip)

### PR DESCRIPTION
**Pull request by rune-scape**
this was supposed to be my fix for this regression: https://github.com/godotengine/godot/issues/92534
it does fix it separately from https://github.com/godotengine/godot/pull/92544

rethinking the changes i made in https://github.com/godotengine/godot/pull/92035 and https://github.com/godotengine/godot/pull/90601

- always default initializes static variables right after compiling a class, safer and less restrictive
- doesn't invalidate script when dependent scripts don't compile/resolve, as this may have caused other (non segfault) problems i didn't imagine
- each `GDScriptFunction` on invalid scripts are protected with a check to see if the script is valid instead of returning too early

this all means that invalid scripts can still have their members safely queried and possibly even native base functions called if there are instances with the invalid script attached
and that a script is not 'invalid' if only a depended script is 'invalid' (as it was before) 

fixes https://github.com/godotengine/godot/issues/92610

### CODE REVIEW

**Comment by CsloudX** [comment]
Thanks, i'm tested with this pull, it fixed https://github.com/godotengine/godot/issues/92610

File: modules/gdscript/gdscript.cpp [hunk 1]
```diff
@@ -904,18 +905,15 @@ void GDScript::unload_static() const {
...
+			HashMap<StringName, GDScriptFunction *>::Iterator E = top->member_functions.find(p_method);
+			if (E) {
+				ERR_FAIL_COND_V_MSG(!E->value->is_static(), Variant(), "Can't call non-static function '" + String(p_method) + "' in script.");
 
-			return E->value->call(nullptr, p_args, p_argcount, r_error);
+				return E->value->call(nullptr, p_args, p_argcount, r_error);
+			}
 		}
 		top = top->_base;
 	}
```
Comment (dalexeev) [inline modules/gdscript/gdscript.cpp:919]: I'm not sure how critical this is, but there may be a situation where the current script is not yet valid, but the base script is already valid. In this case, static function lookup will work inconsistently for "inherited" (actually shadowed) static methods. Until the current script compiles successfully, the lookup will find the base method, and then the current class method.

**Comment by bitwes** [comment]
I tested this PR with [GUT](https://github.com/bitwes/Gut) and it fixed the issue.  

**Minor note**
In an overly complicated scenario (which I can describe if necessary), I had to add an additional check of `is_instance_valid` where I was previously just checking for `null`.  Without this check I get `Attempt to call function 'get_current_test_object' in base 'previously freed' on a null instance`.

For all I know, this could be an issue with GUT that was fixed in Godot, and not a Godot issue.  It's a non issue for GUT, I am just documenting my experience with this PR.

**Review by AThousandShips — APPROVED** [review]
Makes sense and looks good in theory, but TIWAGOS

**Review by akien-mga — APPROVED** [review]
As discussed in the release management meeting yesterday, let's go ahead with that change even though GDScript contributors are not available right now for an in-depth review.

This is a follow up to rune's own changes and was tested by several affected users, so it should be good to go and test more widely in the next beta.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**